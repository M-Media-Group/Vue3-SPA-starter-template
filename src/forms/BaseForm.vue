<script setup lang="ts">
import BaseButton from "@/components/BaseButton.vue";
import { type PropType, nextTick, onMounted, onUpdated, ref } from "vue";
import { navIsLoading } from "@/router";

// Prop of submit text
const props = defineProps({
  submitText: {
    type: String,
    default: "Submit",
  },
  endpoint: {
    type: String,
    required: false,
  },
  clearInputTypesOnFailure: {
    type: Array as PropType<string[]>,
    default: () => ["password"],
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  isLoading: {
    type: Boolean,
    default: false,
  },
  showSubmitButton: {
    type: Boolean,
    default: true,
  },
  autoFocus: {
    type: Boolean,
    default: true,
  },
});

const emit = defineEmits(["submit"]);

const formElement = ref<HTMLFormElement>();

const formIsValid = ref(false);

const checkValidity = () => {
  // For each element in the form, check if it's valid
  callActionsOnAllInputs((element) => {
    // If the element is invalid, add the invalid class
    if (
      !element.validity.valid &&
      !element.validity.valueMissing &&
      !isElementInFocus(element)
    ) {
      setErrorMessageOnElement(element);
    } else {
      clearErrorMessageOnElement(element);
    }
  });

  formIsValid.value = formElement.value?.checkValidity() ?? false;
};

const setErrorMessageOnElement = (element: HTMLInputElement) => {
  //  Update or create a sibling element with the error message
  let errorElement = element.nextElementSibling as HTMLElement;
  if (!errorElement || !errorElement.classList.contains("error")) {
    errorElement = document.createElement("small");
    element.after(errorElement);
  }
  errorElement.innerText = element.validationMessage;
  errorElement.classList.add("error");

  element.insertAdjacentElement("afterend", errorElement);
};

const clearErrorMessageOnElement = (element: HTMLInputElement) => {
  // If the element is valid, remove the invalid class
  const errorElement = element.nextElementSibling as HTMLElement;
  if (errorElement && errorElement.classList.contains("error")) {
    errorElement.remove();
  }
};

const submit = async () => {
  if (
    formIsValid.value &&
    !props.isLoading &&
    !navIsLoading.value &&
    !props.disabled
  ) {
    // setSuccessOnInputs();
    emit("submit");
  }
};

const callActionsOnAllInputs = (
  callback: (element: HTMLInputElement) => void
) => {
  if (!formElement.value?.elements) {
    return;
  }
  const elements = formElement.value.elements;
  for (let i = 0; i < elements.length; i++) {
    const element = elements[i];
    if (element instanceof HTMLInputElement) {
      callback(element);
    }
  }
};

const setInputErrors = (errors: Record<string, string | string[]>) => {
  // For each key in errors, find the input and call setErrorOnInput with the value
  for (const [key, value] of Object.entries(errors)) {
    const input = formElement.value?.elements.namedItem(
      key
    ) as HTMLInputElement;
    if (input) {
      // If the value is an array, join it with a space
      let valueToPass = value as string | string[];

      // If the valueToPass is an array, join it with a space
      if (Array.isArray(valueToPass)) {
        valueToPass = valueToPass.join(" ");
      }

      setErrorOnInput(input, valueToPass);
    }
  }
};

const setErrorOnInput = (
  input: HTMLInputElement,
  error: string,
  report = true
) => {
  //   Use setValidity to set the error message
  input.setCustomValidity(error);
  input.setAttribute("aria-invalid", "true");
  if (report) {
    input.reportValidity();
  }
  setErrorMessageOnElement(input);
};

const setSuccessOnInputs = () => {
  callActionsOnAllInputs((element) => {
    setSuccessOnInput(element);
  });
  // After 5 seconds, clear the success state
  setTimeout(() => {
    removeSuccessOnInputs();
  }, 5000);
};

const setSuccessOnInput = (input: HTMLInputElement) => {
  input.setAttribute("aria-invalid", "false");
};

const removeSuccessOnInputs = () => {
  callActionsOnAllInputs((element) => {
    removeSuccessOnInput(element);
  });
};

const removeSuccessOnInput = (input: HTMLInputElement) => {
  if (input.validity.valid) {
    input.removeAttribute("aria-invalid");
  }
};

const resetCustomValidityOnInput = (input: HTMLInputElement) => {
  input.setCustomValidity("");
  clearErrorMessageOnElement(input);
};

const resetCustomValidityOnInputs = () => {
  // For each element in the form, check if it's valid
  callActionsOnAllInputs((element) => {
    resetCustomValidityOnInput(element);
  });
};

const isElementInFocus = (element: HTMLInputElement) => {
  return document.activeElement === element;
};

// The updated event happens anytime there is input or the slots change
onUpdated(async () => {
  await nextTick();
  checkValidity();
});

onMounted(() => {
  checkValidity();
  // Focus on the first input if the document is not already focused on something
  if (!document.querySelector(":focus") && props.autoFocus) {
    focusOnFirstInput();
  }
});

const focusOnFirstInput = () => {
  const firstInput = formElement.value?.querySelector("input");
  if (firstInput) {
    firstInput.focus();
  }
};

const handleInput = () => {
  resetCustomValidityOnInputs();
  checkValidity();
};

defineExpose({
  checkValidity,
  setInputErrors,
  focusOnFirstInput,
  setSuccessOnInputs,
});
</script>

<template>
  <form
    ref="formElement"
    @input="handleInput"
    @keydown.enter.prevent="submit"
    @submit.prevent="submit"
  >
    <slot></slot>

    <slot
      name="submit"
      :submitText="submitText"
      :submit="submit"
      :disabled="!formIsValid || disabled || isLoading || navIsLoading"
      :isLoading="isLoading"
    >
      <BaseButton
        v-if="showSubmitButton"
        type="submit"
        :disabled="!formIsValid || disabled || isLoading || navIsLoading"
        :aria-busy="isLoading"
      >
        {{ $t(submitText) }}
      </BaseButton>
    </slot>
  </form>
</template>
