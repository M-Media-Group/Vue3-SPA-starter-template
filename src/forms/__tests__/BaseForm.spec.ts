import { describe, expect, it } from "vitest";

import { mount } from "@vue/test-utils";
import BaseForm from "../BaseForm.vue";

describe("Base Form", () => {
  it("renders a form element", () => {
    const wrapper = mount(BaseForm);

    // Expect a form element to exist
    expect(wrapper.find("form").exists()).toBe(true);

    // Expect the form element to have no inputs
    expect(wrapper.find("input").exists()).toBe(false);

    // Expect there to be a submit button
    expect(wrapper.find("button").exists()).toBe(true);

    // The submit button should have a type set to submit
    expect(wrapper.find("button").attributes("type")).toBe("submit");
  });

  it("renders the submit text correctly", () => {
    const wrapper = mount(BaseForm, {
      props: {
        submitText: "Submit",
      },
    });

    // Expect the submit button to have the correct text
    expect(wrapper.find("button").text()).toBe("Submit");
  });

  it("accepts inputs in its default slot", () => {
    const wrapper = mount(BaseForm, {
      slots: {
        default: `
          <input type="text" />
          <input type="text" />
        `,
      },
    });

    // Expect the form element to have two inputs
    expect(wrapper.findAll("input").length).toBe(2);
  });

  it("emits a submit event when submitted", () => {
    const wrapper = mount(BaseForm);
    // Submit the form
    wrapper.find("form").trigger("submit");

    // Expect the form to have emitted a submit event
    expect(wrapper.emitted("submit")).toBeTruthy();
  });

  it("prevents submission when isLoading", () => {
    const wrapper = mount(BaseForm, {
      props: {
        isLoading: true,
      },
    });
    // Submit the form
    wrapper.find("form").trigger("submit");

    // Expect the form to have emitted a submit event
    expect(wrapper.emitted("submit")).toBeFalsy();
  });

  it("emits a submit event when submitted via enter key", () => {
    const wrapper = mount(BaseForm, {
      slots: {
        default: `
          <input type="text" />
        `,
      },
    });
    // Submit the form with a keydown enter event
    wrapper.find("input").trigger("keydown.enter");

    // Expect the form to have emitted a submit event
    expect(wrapper.emitted("submit")).toBeTruthy();
  });

  // There is currently some issue with test-utils not emitting events on clicks, see https://stackoverflow.com/questions/53382235/trigger-form-submit-on-button-click-in-vue-unit-test
  it.skip("emits a submit event when submitted via submit button", () => {
    const wrapper = mount(BaseForm, {
      slots: {
        default: `
          <input type="text" />
        `,
      },
    });
    // Submit the form with a keydown enter event
    wrapper.find("button").trigger("click");

    // Expect the form to have emitted a submit event
    expect(wrapper.emitted("submit")).toBeTruthy();
  });

  it("does not submit if form is invalid", () => {
    const wrapper = mount(BaseForm, {
      slots: {
        default: `
          <input type="text" required />
        `,
      },
    });

    // Expect the submit button to be disabled
    expect(wrapper.find("button").attributes("disabled")).toBeDefined();

    // Submit the form
    wrapper.find("form").trigger("submit");

    // Expect the form to have emitted a submit event
    expect(wrapper.emitted("submit")).toBeFalsy();
  });

  it("disables the submit button when the disabled attribute is passed", () => {
    const wrapper = mount(BaseForm, {
      props: {
        disabled: true,
      },
    });

    // Expect the submit button to be disabled
    expect(wrapper.find("button").attributes("disabled")).toBeDefined();
  });

  it("disables the submit button and sets aria-busy when the isLoading attribute is passed", () => {
    const wrapper = mount(BaseForm, {
      props: {
        isLoading: true,
      },
    });

    // Expect the submit button to be disabled
    expect(wrapper.find("button").attributes("disabled")).toBeDefined();

    // The aria-busy should be true
    expect(wrapper.find("button").attributes("aria-busy")).toBe("true");
  });

  it("is does not disable the submit button by default", () => {
    const wrapper = mount(BaseForm);

    // Expect the submit button to be disabled
    // @todo, according the the HTML spec the attribute shouldnt even exist on the element
    expect(wrapper.find("button").attributes("disabled")).toBeFalsy();
  });

  it("it sets input errors", () => {
    const wrapper = mount(BaseForm, {
      slots: {
        default: `
          <input type="text" name="test" value="test" required />
        `,
      },
    });

    // Call the setInputErrors method on the wrapper
    wrapper.vm.setInputErrors({
      test: ["This is an error"],
    });

    // Expect the input validity to be false
    expect(wrapper.find("input").element.validity.valid).toBe(false);
    expect(wrapper.find("input").attributes("aria-invalid")).toBe("true");
  });

  it("sets success on inputs", () => {
    const wrapper = mount(BaseForm, {
      slots: {
        default: `
          <input type="text" name="test" value="test" required />
        `,
      },
    });

    // Call the setInputSuccess method on the wrapper
    wrapper.vm.setSuccessOnInputs();

    // Expect the input validity to be false
    expect(wrapper.find("input").element.validity.valid).toBe(true);

    // The aria-invalid should be false
    expect(wrapper.find("input").attributes("aria-invalid")).toBe("false");
  });

  it("can focus on first input", () => {
    const wrapper = mount(BaseForm, {
      attachTo: document.body,
      slots: {
        default: `
          <input type="text" name="test" value="test" required />
          <input type="text" name="test2" value="test" required />
          `,
      },
    });

    // Call the focusFirstInput method on the wrapper
    wrapper.vm.focusOnFirstInput();

    // Expect the input to be focused on the first input with the name test
    expect(document.activeElement).toBe(
      wrapper.find("input[name='test']").element
    );
  });

  it("does not show the submit button if the showSubmitButton prop is false", () => {
    const wrapper = mount(BaseForm, {
      props: {
        showSubmitButton: false,
      },
    });

    // Expect the submit button to be disabled
    expect(wrapper.find("button").exists()).toBe(false);
  });

  it("shows a custom button is passed in the submit slot", () => {
    const wrapper = mount(BaseForm, {
      slots: {
        submit: `
          <button type="submit" id="customSubmit">Submit</button>
        `,
      },
    });

    // Expect a customSubmit button to exist
    expect(wrapper.find("#customSubmit").exists()).toBe(true);
  });

  it("shows submitText in the submit slot if no custom button is passed", () => {
    const wrapper = mount(BaseForm, {
      props: {
        submitText: "Test Submit",
      },
      slots: {
        submit: `
          <button type="submit" id="customSubmit">{{submitText}}</button>
        `,
      },
    });

    // Expect the submit button to be disabled
    expect(wrapper.find("button").text()).toBe("Test Submit");
  });

  it("can be submitted with a custom event in the submit slot", async () => {
    const wrapper = mount(BaseForm, {
      slots: {
        submit: `
          <input type="checkbox" id="customSubmit" @change="submit" />
        `,
      },
    });

    // Submit the form by checking the checkbox
    wrapper.find("#customSubmit").trigger("change");

    // Expect the form to have emitted a submit event
    expect(wrapper.emitted("submit")).toBeTruthy();
  });

  // See issue and SO link above for submit with click
  it.skip("submits form if a custom button is passed in the submit slot", () => {
    const wrapper = mount(BaseForm, {
      slots: {
        submit: `
          <button type="submit" id="customSubmit">Submit</button>
        `,
      },
    });

    // Submit the form
    wrapper.find("#customSubmit").trigger("click");

    // Expect the form to have emitted a submit event
    expect(wrapper.emitted("submit")).toBeTruthy();
  });
});
