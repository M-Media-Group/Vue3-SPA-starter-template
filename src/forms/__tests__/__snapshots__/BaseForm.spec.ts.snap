// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Base Form > accepts inputs in its default slot 1`] = `
"<form>
  <!-- @slot This is the default slot for the form. You can use this to add any input or button you want. --><input type="text"><input type="text"><!-- @slot This is the slot for the submit button. You can use this to add a custom submit button or action. --><button disabled="" type="submit" aria-busy="false">Submit</button>
</form>"
`;

exports[`Base Form > renders a form element 1`] = `
"<form>
  <!-- @slot This is the default slot for the form. You can use this to add any input or button you want. -->
  <!-- @slot This is the slot for the submit button. You can use this to add a custom submit button or action. --><button disabled="" type="submit" aria-busy="false">Submit</button>
</form>"
`;

exports[`Base Form > shows a custom button is passed in the submit slot 1`] = `
"<form>
  <!-- @slot This is the default slot for the form. You can use this to add any input or button you want. -->
  <!-- @slot This is the slot for the submit button. You can use this to add a custom submit button or action. --><button type="submit" id="customSubmit">Submit</button>
</form>"
`;
